packages/expo-updates/android/src/main/java/expo/modules/updates/utils/AndroidResourceAssetUtils.kt

```diff
@@ -0,0 +1,113 @@
+
+package expo.modules.updates.utils
+
+import android.annotation.SuppressLint
+import android.content.Context
+import androidx.core.net.toUri
+import expo.modules.core.errors.InvalidArgumentException
+import expo.modules.updates.db.entity.AssetEntity
+import java.io.IOException
+
+/**
+ * Helpers for Android embedded assets and resources
+ */
+internal object AndroidResourceAssetUtils {
+  private const val ANDROID_EMBEDDED_URL_BASE_ASSET = "file:///android_asset/"
+  private const val ANDROID_EMBEDDED_URL_BASE_RESOURCE = "file:///android_res/"
+
+  /**
+   * Create an embedded asset filename in `file:///android_res/` or `file:///android_asset/` format
+   */
+  fun createEmbeddedFilenameForAsset(asset: AssetEntity): String? {
+    val fileExtension = asset.getFileExtension()
+    if (asset.embeddedAssetFilename != null) {
+      return "${ANDROID_EMBEDDED_URL_BASE_ASSET}${asset.embeddedAssetFilename}$fileExtension"
+    }
+    if (asset.resourcesFolder != null && asset.resourcesFilename != null) {
+      return "${ANDROID_EMBEDDED_URL_BASE_RESOURCE}${asset.resourcesFolder}${getDrawableSuffix(asset.scale)}/${asset.resourcesFilename}$fileExtension"
+    }
+    return null
+  }
+
+  /**
+   * Return whether the filePath is an Android asset or resource
+   */
+  fun isAndroidResourceAsset(filePath: String): Boolean {
+    return filePath.startsWith(ANDROID_EMBEDDED_URL_BASE_RESOURCE) ||
+      filePath.startsWith(ANDROID_EMBEDDED_URL_BASE_ASSET)
+  }
+
+  /**
+   * Check a given file name is existed in Android embedded assets
+   */
+  fun isAndroidAssetExisted(context: Context, name: String): Boolean {
+    return try {
+      context.assets.open(name).close()
+      true
+    } catch (e: IOException) {
+      false
+    }
+  }
+
+  /**
+   * Check a given resource folder and filename is existed in Android embedded resources
+   */
+  @SuppressLint("DiscouragedApi")
+  fun isAndroidResourceExisted(context: Context, resourceFolder: String, resourceFilename: String): Boolean {
+    return context.resources.getIdentifier(
+      resourceFilename,
+      resourceFolder,
+      context.packageName
+    ) != 0
+  }
+
+  /**
+   * Check if given filePath matches and is existed in Android embedded assets or resources
+   */
+  fun isAndroidAssetOrResourceExisted(context: Context, filePath: String): Boolean {
+    val (embeddedAssetFilename, resourceFolder, resourceFilename) = parseAndroidResponseAssetFromPath(filePath)
+    if (embeddedAssetFilename != null) {
+      return isAndroidAssetExisted(context, embeddedAssetFilename)
+    }
+    if (resourceFolder != null && resourceFilename != null) {
+      return isAndroidResourceExisted(context, resourceFolder, resourceFilename)
+    }
+    return false
+  }
+
+  /**
+   * Parse a file path to return as Triple<embeddedAssetFilename, resourcesFolder, resourcesFilename>
+   */
+  fun parseAndroidResponseAssetFromPath(filePath: String): Triple<String?, String?, String?> {
```

Comment:
I believe using a data class instead of a Triple is better. It's easy to misunderstand the results of that function now.

Reply:
+1